src/extensions/mvs/components/primitives.ts

```diff
@@ -582,10 +665,103 @@ function addLineMesh(context: PrimitiveBuilderContext, { groups, mesh }: MeshBui
         const count = Math.ceil(dist / (2 * params.dash_length));
         addFixedCountDashedCylinder(mesh, lStart, lEnd, 1.0, count, true, cylinderProps);
     } else {
+        // NOTE: this: use similar
         addSimpleCylinder(mesh, lStart, lEnd, cylinderProps);
     }
 }
 
+function addBoxMesh(context: PrimitiveBuilderContext, { groups, mesh }: MeshBuilderState, node: MVSNode<'primitive'>, params: MVSPrimitiveParams<'box'>, options?: { skipResolvePosition?: boolean }) {
+    // TODO: need this?
+    // if (!options?.skipResolvePosition) {
+    //     // TODO: need this?
+    //     resolveBasePosition(context, params.start, lStart);
+    //     resolveBasePosition(context, params.end, lEnd);
+    // }
+
+    const a = Vec3(), b = Vec3(), c = Vec3(), d = Vec3();
+    // TODO: as_edges, rotation, edge_radius => cage, rotation matrices
+    const { center, extent, box_groups, scaling, as_edges, edge_radius, rotation, translation } = params;
+    const mat4 = Mat4.identity();
+    const triangleCount = 12;
+    // const vertexCount = perforated ? 12 * 3 : 6 * 4;
+    const vertexCount = 6 * 4;
+    const builder = PrimitiveBuilder(triangleCount, vertexCount);
+
+
+    mesh.currentGroup = groups.allocateSingle(node);
+    groups.updateColor(mesh.currentGroup, params.color);
+    // groups.updateTooltip(mesh.currentGroup, params.tooltip);
+
+    // if (params.dash_length) {
+    //     const dist = Vec3.distance(lStart, lEnd);
+    //     const count = Math.ceil(dist / (2 * params.dash_length));
+    //     addFixedCountDashedCylinder(mesh, lStart, lEnd, 1.0, count, true, cylinderProps);
+    // } else {
+    //     // NOTE: this: use similar
+    //     addSimpleCylinder(mesh, lStart, lEnd, cylinderProps);
+    // }
+    // addSimpleCylinder(mesh, lStart, lEnd, cylinderProps);
+
+    // S1
+    Vec3.set(a, 0, 0, 0);
+    Vec3.set(b, 0, 0, 1);
+    Vec3.set(c, 0, 1, 1);
+    Vec3.set(d, 0, 1, 0);
+    builder.addQuad(a, b, c, d);
+
+    // S2
+    Vec3.set(a, 0, 0, 1);
+    Vec3.set(b, 0, 1, 1);
+    Vec3.set(c, 1, 1, 1);
+    Vec3.set(d, 1, 0, 1);
+    builder.addQuad(a, b, c, d);
+
+    // S3
+    Vec3.set(a, 1, 1, 1);
+    Vec3.set(b, 1, 0, 1);
+    Vec3.set(c, 1, 0, 0);
+    Vec3.set(d, 1, 1, 0);
+    builder.addQuad(a, b, c, d);
+
+    // S4
+    Vec3.set(a, 1, 0, 0);
+    Vec3.set(b, 1, 1, 0);
+    Vec3.set(c, 0, 1, 0);
+    Vec3.set(d, 0, 0, 0);
+    builder.addQuad(a, b, c, d);
+
+
+    // TOP
+    Vec3.set(a, 0, 0, 0);
+    Vec3.set(b, 0, 0, 1);
+    Vec3.set(c, 1, 0, 1);
+    Vec3.set(d, 1, 0, 0);
+    builder.addQuad(a, b, c, d);
+
+    // BOTTOM
+    Vec3.set(a, 0, 1, 0);
+    Vec3.set(b, 0, 1, 1);
+    Vec3.set(c, 1, 1, 1);
+    Vec3.set(d, 1, 1, 0);
+    builder.addQuad(a, b, c, d);
```

Comment:
Would `MeshBuilder.addCage` work here? Moreover, you are building a static geometry here transformed by the matrix, you only need to do this once, not separately for each box that will be rendered.

Reply:
> Would `MeshBuilder.addCage` work here? Moreover, you are building a static geometry here transformed by the matrix, you only need to do this once, not separately for each box that will be rendered.

@dsehnal, yes, I have thought about the `MeshBuilder.addCage` as well. To my understanding, it builds the cage, not box, which is required here.

Regarding the static geometry - yes, I am aware of it. As I wrote earlier on, this is hardcoded prototype just to understand how primitive visualisation works under the hood. After fixing the issue with vertex ordering that you have mentioned earlier, I was planning to convert this to an algorithm, similar to [the existing one](https://github.com/molstar/molstar/blob/b4238f574ac3b7a20d0f368812a4237a29d56b8c/src/mol-geo/primitive/box.ts#L18). The key difference should be that it will build the sides of the box based on the user provided parameters (`center`, `extent` etc.), not based on the "hardcoded" `polygon` points. Hope my explanation is clear.

However, based on what you have written, I now understand that [this algorithm](https://github.com/molstar/molstar/blob/b4238f574ac3b7a20d0f368812a4237a29d56b8c/src/mol-geo/primitive/box.ts#L18) can be actually re-used "as is", and will just need to set correct transformation matrix [here](https://github.com/molstar/molstar/blob/e81784cd18f5fbb892c3f44eaaddab1cfc45d3d7/src/extensions/mvs/components/primitives.ts#L684) and apply it [here](https://github.com/molstar/molstar/blob/e81784cd18f5fbb892c3f44eaaddab1cfc45d3d7/src/extensions/mvs/components/primitives.ts#L761C36-L761C40). Did I get you right?

Reply:
> However, based on what you have written, I now understand that [this algorithm](https://github.com/molstar/molstar/blob/b4238f574ac3b7a20d0f368812a4237a29d56b8c/src/mol-geo/primitive/box.ts#L18) can be actually re-used "as is", and will just need to set correct transformation matrix [here](https://github.com/molstar/molstar/blob/e81784cd18f5fbb892c3f44eaaddab1cfc45d3d7/src/extensions/mvs/components/primitives.ts#L684) and apply it [here](https://github.com/molstar/molstar/blob/e81784cd18f5fbb892c3f44eaaddab1cfc45d3d7/src/extensions/mvs/components/primitives.ts#L761C36-L761C40). Did I get you right?

Looks right. Feel free to just export the function. Also, still keep in mind the primitive only needs to be built once.